content/articles/writing-my-first-crate.md

```diff
@@ -0,0 +1,154 @@
+Title: Writing my first Rust crate: jsonwebtoken
+Date: 2015-11-03
+Short_summary: I talk about the experience of writing a library in Rust to handle JWTs 
+Category: Dev
+Authors: Vincent
+
+*Reading time: ~10 minutes.*
+
+After [looking into Rust for webservices]({filename}/articles/rust-for-web-services.md), I concluded that, while it was not mature enough yet in my opinion, the language itself is quite nice and I would be interested in writing more of it. So here it is, my first crate (the name for package in Rust): [jsonwebtoken](https://crates.io/crates/jsonwebtoken).
+
+<!-- PELICAN_END_SUMMARY -->
+
+
+## What are JWTs
+First of all a very quick introduction on JSON Web Token (JWT), if you already know about them you can skip this.
+
+They are a way to transmit data using JSON — hence the name — and are mostly used for APIs and token based authentication. You can read the standard [there](http://self-issued.info/docs/draft-ietf-oauth-json-web-token.html).
+
+Rather than having a session in a database, you would store some non-sensitive data in your token that allow you to identify the user.
+
+Here's what a JWT looks like (with the fields shortened for example sake):
+
+```js
+// header.payload.signature
+eyJhbGciOiJIUzI1Ni.eyJzdWIiOiIxMjM0NTY3ODkwIn0.Rq8IxqeX7eA6GgYxlcHdP
+```
+
+From the code above, you can see that a JWT has 3 parts, separated by a `.`.
+All of those parts are base64 encoded, which means is trivial to decode (remember the non-sensitive data point above?).
+
+
+The first is the header, that typically contains the following:
+```js
+{
+  "alg": "HS256", // which algorithm was used to sign
+  "typ": "JWT"  // actually optional
+}
+```
+
+The second is the claims object, which contains an arbitrary JSON object. There are 
+some reserved claim name such as exp for an expiration timestamp but none of them are
+mandatory. Typically in an app, you would store the user id along as some token metadata such as exp mentioned above for example.
+
+The last part is the signature which is obtained the following way in pseudo-code:
+
+```
+  payload = base64(header) + "." + base64(claims)
+  // "secret" is your secret private key
+  signature = Hmac(payload, "secret", Sha256) 
+```
+
+As you can see it's quite a convenient way to transfer some data and the signature ensures that the payload was not tampered with (there are some issues though, such as [this vulnerability](https://auth0.com/blog/2015/03/31/critical-vulnerabilities-in-json-web-token-libraries/)).
+
+
+## Enter jsonwebtoken
+I was implementing JWT in a Go project using the [jwt-go](https://github.com/dgrijalva/jwt-go) library when I realised a JWT library was something simple enough that I could try in Rust.
+
+I didn't implement the standard to the letter as I wanted to keep it simple for now and not support some part of it (why is `none` a valid value for the alg header?).
```

Comment:
Sentence isn't clear to me: You omitted the none implementation? (if so: good! but unclear from description)